src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp

```diff
@@ -74,31 +412,41 @@ HeapWord* ShenandoahFreeSet::allocate_single(ShenandoahAllocRequest& req, bool&
   // Free set maintains mutator and collector views, and normally they allocate in their views only,
   // unless we special cases for stealing and mixed allocations.
 
+  // Overwrite with non-zero (non-NULL) values only if necessary for allocation bookkeeping.
+
   switch (req.type()) {
     case ShenandoahAllocRequest::_alloc_tlab:
     case ShenandoahAllocRequest::_alloc_shared: {
-
       // Try to allocate in the mutator view
-      for (size_t idx = _mutator_leftmost; idx <= _mutator_rightmost; idx++) {
-        if (is_mutator_free(idx)) {
-          HeapWord* result = try_allocate_in(_heap->get_region(idx), req, in_new_region);
-          if (result != nullptr) {
-            return result;
+      // Allocate within mutator free from high memory to low so as to preserve low memory for humongous allocations
+      if (!_free_sets.is_empty(Mutator)) {
+        // Use signed idx.  Otherwise, loop will never terminate.
```

Comment:
Can you use `ssize_t` for the index and the return values from the methods? (I assume the problem you have in mind arises when `leftmost` below is 0?

Reply:
The problem happens when leftmost is zero.  If we decrement idx beyond zero, we would get MAXINT rather than -1, so the test that idx >= leftmost is always true and loop never terminates...  I think I discovered this the hard way... :(

Reply:
`ssize_t` is signed, unlike `size_t` which is unsigned.

Reply:
Thanks for clarifying.  I did know ssize_t.  I'm adjusting the code.